File: src/database/connector_installations.py
"""Repository for managing connector installations in the control database."""

import json
from datetime import datetime
from enum import Enum
from uuid import UUID

import asyncpg

from src.utils.config import get_control_database_url


class ConnectorType(str, Enum):
    """Valid connector types matching TypeScript ConnectorType enum."""

    SLACK = "slack"
    GITHUB = "github"
    LINEAR = "linear"
    NOTION = "notion"
    GOOGLE_DRIVE = "google_drive"
    GOOGLE_EMAIL = "google_email"
    HUBSPOT = "hubspot"
    SALESFORCE = "salesforce"
    JIRA = "jira"
    CONFLUENCE = "confluence"
    GONG = "gong"
    GATHER = "gather"
    TRELLO = "trello"
    ZENDESK = "zendesk"
    ASANA = "asana"
    INTERCOM = "intercom"
    SNOWFLAKE = "snowflake"
    FIREFLIES = "fireflies"
    ATTIO = "attio"
    PIPEDRIVE = "pipedrive"
    CLICKUP = "clickup"
    PYLON = "pylon"
    MONDAY = "monday"
    FIGMA = "figma"
    POSTHOG = "posthog"
    CANVA = "canva"
    TEAMWORK = "teamwork"


class ConnectorStatus(str, Enum):
    """Valid connector statuses."""

    PENDING = "pending"
    ACTIVE = "active"
    ERROR = "error"
    DISCONNECTED = "disconnected"


class Connector:
    """Connector data model."""

    def __init__(self, row: asyncpg.Record):
        self.id: UUID = row["id"]
        self.tenant_id: str = row["tenant_id"]
        self.type: str = row["type"]
        self.external_id: str = row["external_id"]
        self.external_metadata: dict = row["external_metadata"] or {}
        self.status: str = row["status"]
        self.created_at: datetime = row["created_at"]
        self.updated_at: datetime = row["updated_at"]

    def to_dict(self) -> dict:
        """Convert to dictionary."""
        return {
            "id": str(self.id),
            "tenant_id": self.tenant_id,
            "type": self.type,
            "external_id": self.external_id,
            "external_metadata": self.external_metadata,
            "status": self.status,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }


class ConnectorInstallationsRepository:
    """Repository for connector installation CRUD operations."""

    def __init__(self):
        self.control_db_url = get_control_database_url()

    async def _get_connection(self) -> asyncpg.Connection:
        """Get database connection."""
        return await asyncpg.connect(self.control_db_url)

    async def get_by_id(self, connector_id: UUID) -> Connector | None:
        """Get connector by ID."""
        conn = await self._get_connection()
        try:
            row = await conn.fetchrow(
                """
                SELECT id, tenant_id, type, external_id, external_metadata,
                       status, created_at, updated_at
                FROM connector_installations
                WHERE id = $1
                """,
                connector_id,
            )
            return Connector(row) if row else None
        finally:
            await conn.close()

    async def get_by_tenant(self, tenant_id: str) -> list[Connector]:
        """Get all connectors for a tenant."""
        conn = await self._get_connection()
        try:
            rows = await conn.fetch(
                """
                SELECT id, tenant_id, type, external_id, external_metadata,
                       status, created_at, updated_at
                FROM connector_installations
                WHERE tenant_id = $1
                ORDER BY created_at DESC
                """,
                tenant_id,
            )
            return [Connector(row) for row in rows]
        finally:
            await conn.close()

    async def get_by_tenant_and_type(
        self, tenant_id: str, connector_type: ConnectorType
    ) -> Connector | None:
        """Get connector by tenant and type."""
        conn = await self._get_connection()
        try:
            row = await conn.fetchrow(
                """
                SELECT id, tenant_id, type, external_id, external_metadata,
                       status, created_at, updated_at
                FROM connector_installations
                WHERE tenant_id = $1 AND type = $2
                """,
                tenant_id,
                connector_type,
            )
            return Connector(row) if row else None
        finally:
            await conn.close()

    async def get_by_tenant_type_and_external_id(
        self, tenant_id: str, connector_type: ConnectorType, external_id: str
    ) -> Connector | None:
        """Get connector by tenant, type, and external ID (unique constraint)."""
        conn = await self._get_connection()
        try:
            row = await conn.fetchrow(
                """
                SELECT id, tenant_id, type, external_id, external_metadata,
                       status, created_at, updated_at
                FROM connector_installations
                WHERE tenant_id = $1 AND type = $2 AND external_id = $3
                """,
                tenant_id,
                connector_type,
                external_id,
            )
            return Connector(row) if row else None
        finally:
            await conn.close()

    async def create(
        self,
        tenant_id: str,
        connector_type: ConnectorType,
        external_id: str,
        external_metadata: dict | None = None,
        status: ConnectorStatus = ConnectorStatus.PENDING,
    ) -> Connector:
        """Create a new connector."""
        conn = await self._get_connection()
        try:
            row = await conn.fetchrow(
                """
                INSERT INTO connector_installations (tenant_id, type, external_id, external_metadata, status)
                VALUES ($1, $2, $3, $4, $5)
                RETURNING id, tenant_id, type, external_id, external_metadata,
                          status, created_at, updated_at
                """,
                tenant_id,
                connector_type,
                external_id,
                json.dumps(external_metadata) if external_metadata else None,
                status,
            )
            return Connector(row)
        finally:
            await conn.close()

    async def update_status(self, connector_id: UUID, status: ConnectorStatus) -> None:
        """Update connector status."""
        conn = await self._get_connection()
        try:
            await conn.execute(
                """
                UPDATE connector_installations
                SET status = $1, updated_at = NOW()
                WHERE id = $2
                """,
                status,
                connector_id,
            )
        finally:
            await conn.close()

    async def update_metadata(self, connector_id: UUID, metadata: dict) -> None:
        """Update connector metadata."""
        conn = await self._get_connection()
        try:
            await conn.execute(
                """
                UPDATE connector_installations
                SET external_metadata = $1, updated_at = NOW()
                WHERE id = $2
                """,
                json.dumps(metadata) if metadata else None,
                connector_id,
            )
        finally:
            await conn.close()

    async def delete(self, connector_id: UUID) -> None:
        """Delete a connector (hard delete)."""
        conn = await self._get_connection()
        try:
            await conn.execute(
                """
                DELETE FROM connector_installations
                WHERE id = $1
                """,
                connector_id,
            )
        finally:
            await conn.close()

    async def mark_disconnected(self, connector_id: UUID) -> None:
        """Mark connector as disconnected (soft delete)."""
        await self.update_status(connector_id, ConnectorStatus.DISCONNECTED)

    async def get_by_type_and_external_id(
        self, connector_type: ConnectorType, external_id: str, exclude_disconnected: bool = True
    ) -> Connector | None:
        """Get connector by type and external_id (for webhook routing without tenant_id)."""
        conn = await self._get_connection()
        try:
            query = """
                SELECT id, tenant_id, type, external_id, external_metadata,
                       status, created_at, updated_at
                FROM connector_installations
                WHERE type = $1 AND external_id = $2
            """
            if exclude_disconnected:
                query += " AND status != 'disconnected'"

            row = await conn.fetchrow(query, connector_type, external_id)
            return Connector(row) if row else None
        finally:
            await conn.close()

    async def get_active_tenant_ids_by_type(self, connector_type: str) -> list[str]:
        """Get all tenant IDs that have an active connector of the given type.

        Only returns provisioned tenants with connectors that are not 'disconnected'.
        """
        conn = await self._get_connection()
        try:
            rows = await conn.fetch(
                """
                SELECT DISTINCT ci.tenant_id
                FROM connector_installations ci
                JOIN tenants t ON ci.tenant_id = t.id
                WHERE ci.type = $1
                  AND ci.status = 'active'
                  AND t.state = 'provisioned'
                """,
                connector_type,
            )
            return [row["tenant_id"] for row in rows]
        finally:
            await conn.close()

    async def get_all_tenant_ids_with_connectors(self) -> list[str]:
        """Get all tenant IDs that have any connector installations.

        Only returns provisioned tenants.
        """
        conn = await self._get_connection()
        try:
            rows = await conn.fetch(
                """
                SELECT DISTINCT ci.tenant_id
                FROM connector_installations ci
                JOIN tenants t ON ci.tenant_id = t.id
                WHERE t.state = 'provisioned'
                ORDER BY ci.tenant_id
                """,
            )
            return [row["tenant_id"] for row in rows]
        finally:
            await conn.close()

    async def get_figma_connector_by_team_id(
        self, team_id: str, exclude_disconnected: bool = True
    ) -> Connector | None:
        """Get Figma connector by team_id (searches external_metadata.synced_team_ids).

        Figma webhooks include team_id, which we need to map to a tenant.
        The team_id is stored in external_metadata.synced_team_ids array.
        """
        conn = await self._get_connection()
        try:
            # Use the ? operator to check if a JSONB array contains a string element
            query = """
                SELECT id, tenant_id, type, external_id, external_metadata,
                       status, created_at, updated_at
                FROM connector_installations
                WHERE type = 'figma'
                  AND (
                    external_metadata->'synced_team_ids' ? $1
                    OR external_metadata->'selected_team_ids' ? $1
                  )
            """
            if exclude_disconnected:
                query += " AND status != 'disconnected'"

            row = await conn.fetchrow(query, team_id)
            return Connector(row) if row else None
        finally:
            await conn.close()

File: js-services/admin-backend/src/dal/connector-installations.ts
/**
 * Repository for managing connector installations in the control database.
 */

import { Pool } from 'pg';
import { getControlDbPool } from '../control-db.js';
import { logger } from '../utils/logger.js';
import {
  Connector,
  ConnectorStatus,
  ConnectorType,
  CreateConnectorData,
} from '../types/connector.js';

export class ConnectorInstallationsRepository {
  private pool: Pool | null;

  constructor() {
    this.pool = getControlDbPool();
  }

  /**
   * Get connector by ID
   */
  async getById(id: string): Promise<Connector | null> {
    if (!this.pool) {
      logger.error('Control database not available');
      return null;
    }

    try {
      const result = await this.pool.query(
        `SELECT id, tenant_id, type, external_id, external_metadata,
                status, created_at, updated_at
         FROM connector_installations
         WHERE id = $1`,
        [id]
      );

      if (result.rows.length === 0) {
        return null;
      }

      return this.mapRow(result.rows[0]);
    } catch (error) {
      logger.error('Error getting connector by ID', { id, error });
      throw error;
    }
  }

  /**
   * Get all connectors for a tenant
   */
  async getByTenant(tenantId: string): Promise<Connector[]> {
    if (!this.pool) {
      logger.error('Control database not available');
      return [];
    }

    try {
      const result = await this.pool.query(
        `SELECT id, tenant_id, type, external_id, external_metadata,
                status, created_at, updated_at
         FROM connector_installations
         WHERE tenant_id = $1
         ORDER BY created_at DESC`,
        [tenantId]
      );

      return result.rows.map((row) => this.mapRow(row));
    } catch (error) {
      logger.error('Error getting connectors by tenant', { tenantId, error });
      throw error;
    }
  }

  /**
   * Get connector by tenant and type
   * Returns only active (non-disconnected) connectors, ordered by most recent first
   */
  async getByTenantAndType(tenantId: string, type: ConnectorType): Promise<Connector | null> {
    if (!this.pool) {
      logger.error('Control database not available');
      return null;
    }

    try {
      const result = await this.pool.query(
        `SELECT id, tenant_id, type, external_id, external_metadata,
                status, created_at, updated_at
         FROM connector_installations
         WHERE tenant_id = $1 AND type = $2 AND status != 'disconnected'
         ORDER BY created_at DESC
         LIMIT 1`,
        [tenantId, type]
      );

      if (result.rows.length === 0) {
        return null;
      }

      return this.mapRow(result.rows[0]);
    } catch (error) {
      logger.error('Error getting connector by tenant and type', { tenantId, type, error });
      throw error;
    }
  }

  /**
   * Get disconnected connector by tenant and type
   * Used to retrieve metadata from previously disconnected connectors (e.g., for reconnection scenarios)
   */
  async getDisconnectedByTenantAndType(
    tenantId: string,
    type: ConnectorType
  ): Promise<Connector | null> {
    if (!this.pool) {
      logger.error('Control database not available');
      return null;
    }

    try {
      const result = await this.pool.query(
        `SELECT id, tenant_id, type, external_id, external_metadata,
                status, created_at, updated_at
         FROM connector_installations
         WHERE tenant_id = $1 AND type = $2 AND status = 'disconnected'
         ORDER BY updated_at DESC
         LIMIT 1`,
        [tenantId, type]
      );

      if (result.rows.length === 0) {
        return null;
      }

      return this.mapRow(result.rows[0]);
    } catch (error) {
      logger.error('Error getting disconnected connector by tenant and type', {
        tenantId,
        type,
        error,
      });
      return null;
    }
  }

  /**
   * Get connector by tenant, type, and external ID (unique constraint)
   */
  async getByTenantTypeAndExternalId(
    tenantId: string,
    type: ConnectorType,
    externalId: string
  ): Promise<Connector | null> {
    if (!this.pool) {
      logger.error('Control database not available');
      return null;
    }

    try {
      const result = await this.pool.query(
        `SELECT id, tenant_id, type, external_id, external_metadata,
                status, created_at, updated_at
         FROM connector_installations
         WHERE tenant_id = $1 AND type = $2 AND external_id = $3`,
        [tenantId, type, externalId]
      );

      if (result.rows.length === 0) {
        return null;
      }

      return this.mapRow(result.rows[0]);
    } catch (error) {
      logger.error('Error getting connector by tenant, type, and external ID', {
        tenantId,
        type,
        externalId,
        error,
      });
      throw error;
    }
  }

  /**
   * Create a new connector
   */
  async create(data: CreateConnectorData): Promise<Connector> {
    if (!this.pool) {
      throw new Error('Control database not available');
    }

    try {
      const result = await this.pool.query(
        `INSERT INTO connector_installations (tenant_id, type, external_id, external_metadata, status)
         VALUES ($1, $2, $3, $4, $5)
         RETURNING id, tenant_id, type, external_id, external_metadata,
                   status, created_at, updated_at`,
        [
          data.tenant_id,
          data.type,
          data.external_id,
          JSON.stringify(data.external_metadata || {}),
          data.status || 'pending',
        ]
      );

      return this.mapRow(result.rows[0]);
    } catch (error) {
      logger.error('Error creating connector', { data, error });
      throw error;
    }
  }

  /**
   * Update connector status
   */
  async updateStatus(id: string, status: ConnectorStatus): Promise<void> {
    if (!this.pool) {
      throw new Error('Control database not available');
    }

    try {
      await this.pool.query(
        `UPDATE connector_installations
         SET status = $1, updated_at = NOW()
         WHERE id = $2`,
        [status, id]
      );
    } catch (error) {
      logger.error('Error updating connector status', { id, status, error });
      throw error;
    }
  }

  /**
   * Update connector metadata
   */
  async updateMetadata(id: string, metadata: Record<string, unknown>): Promise<void> {
    if (!this.pool) {
      throw new Error('Control database not available');
    }

    try {
      await this.pool.query(
        `UPDATE connector_installations
         SET external_metadata = $1, updated_at = NOW()
         WHERE id = $2`,
        [JSON.stringify(metadata), id]
      );
    } catch (error) {
      logger.error('Error updating connector metadata', { id, error });
      throw error;
    }
  }

  /**
   * Delete a connector (hard delete)
   */
  async delete(id: string): Promise<void> {
    if (!this.pool) {
      throw new Error('Control database not available');
    }

    try {
      await this.pool.query('DELETE FROM connector_installations WHERE id = $1', [id]);
    } catch (error) {
      logger.error('Error deleting connector', { id, error });
      throw error;
    }
  }

  /**
   * Mark connector as disconnected (soft delete)
   */
  async markDisconnected(id: string): Promise<void> {
    await this.updateStatus(id, 'disconnected');
  }

  /**
   * Get connector by type and external ID (for webhook routing without tenant_id)
   * Returns only active (non-disconnected) connectors
   */
  async getByTypeAndExternalId(type: ConnectorType, externalId: string): Promise<Connector | null> {
    if (!this.pool) {
      logger.error('Control database not available');
      return null;
    }

    try {
      const result = await this.pool.query(
        `SELECT id, tenant_id, type, external_id, external_metadata,
                status, created_at, updated_at
         FROM connector_installations
         WHERE type = $1 AND external_id = $2 AND status != 'disconnected'`,
        [type, externalId]
      );

      if (result.rows.length === 0) {
        return null;
      }

      return this.mapRow(result.rows[0]);
    } catch (error) {
      logger.error('Error getting connector by type and external ID', { type, externalId, error });
      throw error;
    }
  }

  /**
   * Get all connectors of a specific type (for cron jobs)
   * Returns only active (non-disconnected) connectors
   */
  async getAllByType(type: ConnectorType): Promise<Connector[]> {
    if (!this.pool) {
      logger.error('Control database not available');
      return [];
    }

    try {
      const result = await this.pool.query(
        `SELECT id, tenant_id, type, external_id, external_metadata,
                status, created_at, updated_at
         FROM connector_installations
         WHERE type = $1 AND status != 'disconnected'`,
        [type]
      );
      return result.rows.map((row) => this.mapRow(row));
    } catch (error) {
      logger.error('Error getting all connectors by type', { type, error });
      throw error;
    }
  }

  /**
   * Map database row to Connector object
   */
  private mapRow(row: Record<string, unknown>): Connector {
    return {
      id: row.id as string,
      tenant_id: row.tenant_id as string,
      type: row.type as ConnectorType,
      external_id: row.external_id as string,
      external_metadata: (row.external_metadata as Record<string, unknown>) || {},
      status: row.status as ConnectorStatus,
      created_at: (row.created_at as Date).toISOString(),
      updated_at: (row.updated_at as Date).toISOString(),
    };
  }
}
